### PR TITLE
Implement configurable ExecutorService creation for Alluxio processes

### LINKED ISSUE #14326: Configurable ExecutorService provider for Alluxio processes
**Is your feature request related to a problem? Please describe.**
Main executors that are used by master/worker processes are not configurable.
This leaves us with custom code changes to mitigate issues.

**Describe the solution you'd like**
By leveraging propertry-key templates, we could easily provide common infra to configure executor-services for Alluxio nodes.
There are not really many options when it comes to providing executor-services to gRPC servers:
- FJP
- Fixed Thread Pool
- Cached Thread Pool
For the last two, we could also provide support for several internal queue implementations. (LinkedBlocking, SynchronousBlocking etc...)

**Urgency**
It would be very nice to have before the next release.

### PR DESCRIPTION
**Pull request by ggezer**
Fix #14326

### CODE REVIEW

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 1]
```diff
@@ -2737,53 +2737,90 @@ public String toString() {
           .setScope(Scope.MASTER)
           .setConsistencyCheckLevel(ConsistencyCheckLevel.WARN)
           .build();
-  // In Java8 in container environment Runtime.availableProcessors() always returns 1,
-  // which is not the actual number of cpus, so we set a safe default value 8.
-  public static final PropertyKey MASTER_RPC_EXECUTOR_PARALLELISM =
```
Comment (yuzhu) [inline core/common/src/main/java/alluxio/conf/PropertyKey.java:2742]: can we keep these around and set the default FJP or TPE parallelism to this for better back compat?

File: core/server/master/src/test/java/alluxio/master/AlluxioMasterProcessTest.java [hunk 2]
```diff
@@ -171,32 +171,35 @@ public void restoreFromBackupLocal() throws Exception {
 
   @Test
   public void startZeroParallelism() {
-    ServerConfiguration.set(PropertyKey.MASTER_RPC_EXECUTOR_PARALLELISM, "0");
+    ServerConfiguration.set(PropertyKey.MASTER_RPC_EXECUTOR_TYPE, "FJP");
```
Comment (yuzhu) [inline core/server/master/src/test/java/alluxio/master/AlluxioMasterProcessTest.java:174]: can we also add some sanity test for TPE?

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 3]
```diff
@@ -2737,53 +2737,90 @@ public String toString() {
...
-          .setDefaultValue(1)
-          .setDescription("the minimum allowed number of core threads not blocked. "
-              + "To ensure progress, when too few unblocked threads exist and unexecuted tasks may "
-              + "exist, new threads are constructed up to the value of "
-              + Name.MASTER_RPC_EXECUTOR_MAX_POOL_SIZE
-              + ". A value of 1 ensures liveness. A larger value might improve "
-              + "throughput but might also increase overhead.")
+  public static final PropertyKey MASTER_RPC_EXECUTOR_TYPE =
+      new Builder(Name.MASTER_RPC_EXECUTOR_TYPE)
+          .setDefaultValue("TPE")
```
Comment (yuzhu) [inline core/common/src/main/java/alluxio/conf/PropertyKey.java:2742]: can we use full class name instead of TPE and FJP?

**Review by yuzhu — CHANGES_REQUESTED** [review]
thanks @ggezer , some early comments

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 4]
```diff
@@ -2737,53 +2737,90 @@ public String toString() {
...
           .setConsistencyCheckLevel(ConsistencyCheckLevel.WARN)
           .setScope(Scope.MASTER)
           .build();
+  public static final PropertyKey MASTER_RPC_EXECUTOR_TPE_QUEUE_TYPE =
+      new Builder(Name.MASTER_RPC_EXECUTOR_TPE_QUEUE_TYPE)
+          .setDefaultValue("LINKED_BLOCKING_QUEUE")
+          .setDescription(String.format(
+              "This property is effective when %s is set to TPE. "
+                  + "It specifies the internal task queue that's used by RPC ExecutorService. "
+                  + "Supported values are: LINKED_BLOCKING_QUEUE, LINKED_BLOCKING_QUEUE_WITH_CAP, "
```
Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/conf/PropertyKey.java:2778]: If you have an `enum TpeType` you may do `TpeType.values()` here.

File: core/server/common/src/main/java/alluxio/master/AlluxioExecutorService.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+  /**
+   * @return the current RPC queue size
+   */
+  public long getRpcQueueLength() {
+    if (mExecutor instanceof ThreadPoolExecutor) {
+      return ((ThreadPoolExecutor) mExecutor).getQueue().size();
+    } else if (mExecutor instanceof ForkJoinPool) {
+      return ((ForkJoinPool) mExecutor).getQueuedSubmissionCount();
+    } else {
+      throw new IllegalArgumentException("Executor is not supported.");
```
Comment (jiacheliu3) [inline core/server/common/src/main/java/alluxio/master/AlluxioExecutorService.java:50]: print the `Executor.getClassName()` here so we know what other types people may want?

File: job/server/src/main/java/alluxio/master/job/plan/PlanTracker.java [hunk 6]
```diff
@@ -105,8 +105,8 @@ public PlanTracker(long capacity, long retentionMs,
     Preconditions.checkArgument(retentionMs >= 0);
     mRetentionMs = retentionMs;
     mMaxJobPurgeCount = maxJobPurgeCount <= 0 ? Long.MAX_VALUE : maxJobPurgeCount;
-    mCoordinators = new ConcurrentHashMap<>(0,
-        0.95f, ServerConfiguration.getInt(PropertyKey.MASTER_RPC_EXECUTOR_PARALLELISM));
+    mCoordinators = new ConcurrentHashMap<>(0, 0.95f,
```
Comment (jiacheliu3) [inline job/server/src/main/java/alluxio/master/job/plan/PlanTracker.java:108]: Move the comment/description from the property key here so we keep why the `Math(8, 2*nCPU)` here?

File: core/server/common/src/main/java/alluxio/master/ExecutorServiceBuilder.java [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+   * @param executorHost Where the executor is needed
+   * @return instance of {@link ExecutorService}
+   */
+  public static AlluxioExecutorService buildExecutorService(RpcExecutorHost executorHost) {
+    // Get executor type for given host.
+    RpcExecutorType executorType = ServerConfiguration.getEnum(
+        PropertyKey.Template.RPC_EXECUTOR_TYPE.format(executorHost.toString()),
+        RpcExecutorType.class);
+    // Build thread name format.
+    String threadNameFormat = executorHost + "-rpc-executor-thread-%d";
```
Comment (jiacheliu3) [inline core/server/common/src/main/java/alluxio/master/ExecutorServiceBuilder.java:134]: Does it make sense to reflect the threadpool type in the thread names too?

File: core/server/common/src/main/java/alluxio/master/ExecutorServiceBuilder.java [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+          queue = new LinkedBlockingQueue<>(maxPoolSize);
+          break;
+        case ARRAY_BLOCKING_QUEUE:
+          queue = new ArrayBlockingQueue<>(maxPoolSize);
+          break;
+        case SYNCHRONOUS_BLOCKING_QUEUE:
+          queue = new SynchronousQueue<>();
+          break;
+        default:
+          throw new IllegalArgumentException();
```
Comment (jiacheliu3) [inline core/server/common/src/main/java/alluxio/master/ExecutorServiceBuilder.java:210]: Add a message here on what the `queueType` is?

File: core/server/common/src/test/java/alluxio/master/ExecutorServiceBuilderTest.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master;
+
+import alluxio.conf.PropertyKey;
+import alluxio.conf.ServerConfiguration;
+import org.junit.Before;
+import org.junit.Rule;
```
Comment (jiacheliu3) [inline core/server/common/src/test/java/alluxio/master/ExecutorServiceBuilderTest.java:17]: oops checkstyle will complain

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
Thanks for the improvement! Overall this looks great, but I have some high level questions.

I was imagining we do this in a way similar to how we specify `alluxio.worker.allocator.class`. There's an interface for an `ExecutorFactory` and we have `FJPFactory` and `FixedThreadPoolFactor` and the user can write their own factories. This will probably make the configuration easier because the user will go to that class and see how the code(property keys) apply. 

However the current approach also looks good. Partly that's because the TPE constructor only has 7 parameters. Not toooo hard to wrap in property keys.

Just raising an alternative for discussion. Either looks good to me. What do people think?

**Comment by codings-dan** [comment]
@ggezer hi Gorkturk，I would like to know why we chose to add the option of Thread Pool. Compared with FrokJoin Pool, what are its advantages? Are there any performance test results?  thx!

**Comment by ggezer** [comment]
> @ggezer hi Gorkturk，I would like to know why we chose to add the option of Thread Pool. Compared with FrokJoin Pool, what are its advantages? Are there any performance test results? thx!

Hey. With ThreadPoolExecutor we can avoid pitfalls of FJP. FJP is throughout optimized and assumes short execution times. When a task is expected to run for a long time, it should notify FJP via `ForkJoinPool.managedBlock()`. When this is not done, FJP has trouble scaling up its internal thread count. So you could start seeing issues like; worker heartbeats not going through, RPCs not being served etc...

File: job/server/src/main/java/alluxio/master/job/plan/PlanTracker.java (hunk 6 above)
Comment (ggezer) [reply job/server/src/main/java/alluxio/master/job/plan/PlanTracker.java:108]: @yuzhu Why did we relied on this property in the first place?

File: core/common/src/main/java/alluxio/conf/PropertyKey.java (hunk 4 above)
Comment (ggezer) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java:2778]: Can't do unless I move interfaces down to core/common.

File: core/server/master/src/test/java/alluxio/master/AlluxioMasterProcessTest.java (hunk 2 above)
Comment (ggezer) [reply core/server/master/src/test/java/alluxio/master/AlluxioMasterProcessTest.java]: Al those that are asserted are tested already. (keepalive etc is effective for both pools)

File: core/common/src/main/java/alluxio/conf/PropertyKey.java (hunk 1 above)
Comment (ggezer) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java:2742]: I set aliases for old FJP related settings. However they won't take effect due to switching to TPE by default.
Rest are untouched already (core/max pool etc...)

Beyond that, any custom configuration that touches parallelism is most probably wrong.

**Comment by ggezer** [comment]
@yuzhu @jiacheliu3 Now that we have an infra, what do you guys think about introducing keys for job-master and worker processes?

File: core/common/src/main/java/alluxio/conf/PropertyKey.java (hunk 4 above)
Comment (jiacheliu3) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java:2778]: right....

File: job/server/src/main/java/alluxio/master/job/plan/PlanTracker.java (hunk 6 above)
Comment (yuzhu) [reply job/server/src/main/java/alluxio/master/job/plan/PlanTracker.java:107]: Should really make a new property, but everyone hates that..

**Review by yuzhu — APPROVED** [review]

**Comment by yuzhu** [comment]
Lgtm if build and test passes

**Review by jiacheliu3 — APPROVED** [review]
LGTM, thanks!

**Comment by Xenorith** [comment]
alluxio-bot, merge this please

**Comment by Xenorith** [comment]
alluxio-bot, cherry-pick this to branch-2.7 please

**Pull request merged**